### PR TITLE
com.chipmuenk.pyfda

### PR DESCRIPTION
**Pull request by Murmele**
### Please confirm your submission meets all the criteria

- [X] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [X] My pull request follows the instructions at [App Submission][submission].
- [X] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [X] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [X] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [X] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [X] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

**Comment by Murmele** [comment]
bot build com.github.chipmuenk.pyfda

**Comment by Murmele** [comment]
bot build com.github.chipmuenk.pyfda

**Comment by Murmele** [comment]
bot, build com.github.chipmuenk.pyfda

**Comment by Murmele** [comment]
bot, build com.github.chipmuenk.pyfda

**Comment by Murmele** [comment]
bot, build com.github.chipmuenk.pyfda

**Comment by Murmele** [comment]
bot, build com.github.chipmuenk.pyfda

File: com.github.chipmuenk.pyfda.yaml [hunk 1]
```diff
@@ -0,0 +1,58 @@
+app-id: com.github.chipmuenk.pyFDA
+runtime: org.kde.Platform
+runtime-version: '5.12'
```
Comment (hfiguiere) [inline com.github.chipmuenk.pyfda.yaml:3]: this runtime is EOL.
```suggestion
runtime-version: '5.15-21.08'
```

File: com.github.chipmuenk.pyfda.yaml [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+sdk: org.kde.Sdk
+command: pyfdax
+finish-args:
+- "--socket=wayland"
+- "--socket=x11"
+- "--device=dri"
+- "--filesystem=xdg-run/dconf"
+- "--filesystem=~/.config/dconf:ro"
+- "--talk-name=ca.desrt.dconf"
+- "--env=DCONF_USER_CONFIG_DIR=.config/dconf"
```
Comment (hfiguiere) [inline com.github.chipmuenk.pyfda.yaml:13]:
```suggestion
```

With recent runtimes, this is not necessary.

File: com.github.chipmuenk.pyfda.yaml [hunk 3]
```diff
@@ -0,0 +1,58 @@
+app-id: com.github.chipmuenk.pyFDA
+runtime: org.kde.Platform
+runtime-version: '5.12'
+sdk: org.kde.Sdk
+command: pyfdax
+finish-args:
+- "--socket=wayland"
+- "--socket=x11"
```
Comment (hfiguiere) [inline com.github.chipmuenk.pyfda.yaml:8]:
```suggestion
- "--socket=fallback-x11"
```

assuming proper wayland support, otherwise drop the wayland socket.

File: com.github.chipmuenk.pyfda.yaml [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+- "--filesystem=home"
+rename-icon: pyfda_icon #Image will renamed to match the app-id konvention
+rename-appdata-file: pyFDA.appdata.xml
+rename-desktop-file: pyFDA.desktop
+modules:
+- name: PyQt5
+  cleanup:
+  - "/bin/sip"
+  - "/include"
+  - "/lib/python3.7/site-packages/*.pyi"
```
Comment (hfiguiere) [inline com.github.chipmuenk.pyfda.yaml]: the latest runtime use python 3.9 so this will have to change

File: com.github.chipmuenk.pyfda.yaml [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  modules:    
+  - python3-modules.json # include dependencies
+
+- name: pyFDA
+  buildsystem: simple
+  build-commands:
+  - python3 setup.py install --prefix=/app
+  - desktop-file-install --dir=/app/share/applications ressource/mime/applications/pyFDA.desktop
+  - install -Dm755 ressource/pyFDA.appdata.xml /app/share/appdata/pyFDA.appdata.xml
+  - install -Dm755 docs/source/img/pyfda_icon.png /app/share/icons/hicolor/100x100/apps/pyfda_icon.png
```
Comment (hfiguiere) [inline com.github.chipmuenk.pyfda.yaml:28]: 100x100 isn't a valid icon size.

File: com.github.chipmuenk.pyfda.yaml [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  - install -Dm755 docs/source/img/pyfda_icon.svg /app/share/icons/hicolor/scalable/apps/pyfda_icon.svg
+  sources:
+  - type: git
+    branch: develop
```
Comment (hfiguiere) [inline com.github.chipmuenk.pyfda.yaml:32]: you need an explicit commit.

File: com.github.chipmuenk.pyfda.yaml [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+- "--env=DCONF_USER_CONFIG_DIR=.config/dconf"
+- "--own-name=com.github.chipmuenk.pyFDA"
+- "--filesystem=home"
+rename-icon: pyfda_icon #Image will renamed to match the app-id konvention
+rename-appdata-file: pyFDA.appdata.xml
+rename-desktop-file: pyFDA.desktop
+modules: 
+  - python3-modules.json # include dependencies
+
+- name: pyFDA
```
Comment (hfiguiere) [inline com.github.chipmuenk.pyfda.yaml:22]: this whole block must be indented by two space aligned with line 20.

**Review by hfiguiere — CHANGES_REQUESTED** [review]

File: com.github.chipmuenk.pyfda.yaml (hunk 6 above)
Comment (Murmele) [reply com.github.chipmuenk.pyfda.yaml:28]: done

File: com.github.chipmuenk.pyfda.yaml (hunk 4 above)
Comment (Murmele) [reply com.github.chipmuenk.pyfda.yaml]: done

**Comment by Murmele** [comment]
bot, build com.github.chipmuenk.pyfda

File: com.github.chipmuenk.pyfda.yaml (hunk 4 above)
Comment (Murmele) [reply com.github.chipmuenk.pyfda.yaml]: not relevant anymore

File: com.github.chipmuenk.pyfda.yaml (hunk 5 above)
Comment (Murmele) [reply com.github.chipmuenk.pyfda.yaml]: done

File: python3-modules.json [hunk 8]
```diff
@@ -0,0 +1,221 @@
+{
+    "name": "python3-modules",
```
Comment (hfiguiere) [inline python3-modules.json:2]: there are many packages listed more than once in that manifest.

And I'm pretty sure pyqt5 will not build. Why did you change the way it was before?

**Review by hfiguiere — CHANGES_REQUESTED** [review]

**Comment by Murmele** [comment]
bot, build com.github.chipmuenk.pyfda

**Comment by Murmele** [comment]
bot, stop

**Comment by Murmele** [comment]
bot, abort com.github.chipmuenk.pyfda.

File: python3-modules.json (hunk 8 above)
Comment (Murmele) [reply python3-modules.json:2]: pyqt5 is anymore available on riverbankcomputing

**Comment by Murmele** [comment]
bot, build com.github.chipmuenk.pyfda

**Comment by barthalion** [comment]
bot, build com.github.chipmuenk.pyFDA

**Comment by barthalion** [comment]
The manifest filename has to match the ID.

**Comment by Murmele** [comment]
bot, build com.github.chipmuenk.pyfda

**Comment by Murmele** [comment]
bot, build com.github.chipmuenk.pyFDA

**Comment by Murmele** [comment]
bot, build com.github.chipmuenk.pyfda

**Comment by Murmele** [comment]
bot, build com.github.chipmuenk.pyfda

**Comment by Murmele** [comment]
bot, build com.github.chipmuenk.pyfda

**Comment by hfiguiere** [comment]
```
builddir/files/share/appdata/com.github.chipmuenk.pyfda.appdata.xml: FAILED:
• file-invalid          : <screenshot> failed to load [https://github.com/chipmuenk/pyfda/tree/develop/img/pyfda_screenshot_3.png]
• file-invalid          : <screenshot> failed to load [https://github.com/chipmuenk/pyfda/tree/develop/img/pyfda_screenshot_3d_4.png]
• file-invalid          : <screenshot> failed to load [https://github.com/chipmuenk/pyfda/tree/develop/img/pyfda_screenshot_hn.png]
• file-invalid          : <screenshot> failed to load [https://github.com/chipmuenk/pyfda/tree/develop/img/pyfda_scr_shot_baq_impz.png]
• file-invalid          : <screenshot> failed to load [https://github.com/chipmuenk/pyfda/tree/develop/img/pyfda_screenshot_3d_3.png]
• file-invalid          : <screenshot> failed to load [https://github.com/chipmuenk/pyfda/tree/develop/img/pyfda_screenshot_pz.png]
• file-invalid          : <screenshot> failed to load [https://github.com/chipmuenk/pyfda/tree/develop/img/pyfda_screenshot_spec_error.png]
• file-invalid          : <screenshot> failed to load [https://github.com/chipmuenk/pyfda/tree/develop/img/pyfda_scr_shot_3d5_info.png]
Validation of files failed
```

File: com.github.chipmuenk.pyfda.yaml [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+command: pyfdax
+finish-args:
+- "--socket=wayland"
+- "--socket=fallback-x11"
+- "--device=dri"
+- "--own-name=com.github.chipmuenk.pyfda"
+- "--filesystem=home"
+rename-icon: pyfda_icon #Image will renamed to match the app-id konvention
+rename-appdata-file: pyfda.appdata.xml
+rename-desktop-file: pyfda.desktop
```
Comment (hfiguiere) [inline com.github.chipmuenk.pyfda.yaml:16]: these should be above the finish-args (as a common practice)

**Review by hfiguiere — CHANGES_REQUESTED** [review]
also I think most of the python dependencies list could be generated with "flatpak-builder-tools". (you don't need to change that, I'm just suggesting)

File: com.github.chipmuenk.pyfda.yaml (hunk 9 above)
Comment (Murmele) [reply com.github.chipmuenk.pyfda.yaml:16]: done

File: com.github.chipmuenk.pyfda.yaml [hunk 10]
```diff
@@ -0,0 +1,258 @@
+app-id: com.github.chipmuenk.pyfda
+runtime: org.kde.Platform
+runtime-version: '5.15-21.08'
+sdk: org.kde.Sdk
+# when the runtime version changes, check if python version changed, because in pyqt5 the
+# python version is hardcoded!
+command: pyfdax
+finish-args:
+- "--socket=wayland"
```
[the remark this answers is not in the document]
Comment (Murmele) [reply com.github.chipmuenk.pyfda.yaml:9]: done

**Comment by Murmele** [comment]
> https://github.com/chipmuenk/pyfda/tree/develop/img/pyfda_screenshot_3.png

When copying the url, the images can be opened

**Comment by hfiguiere** [comment]
> > https://github.com/chipmuenk/pyfda/tree/develop/img/pyfda_screenshot_3.png
> 
> When copying the url, the images can be opened

No. what you a seeing is github page, not the image itself. You want the URL behind the [Download] button.

**Comment by Murmele** [comment]
> > > https://github.com/chipmuenk/pyfda/tree/develop/img/pyfda_screenshot_3.png
> > 
> > 
> > When copying the url, the images can be opened
> 
> No. what You are seeing is github page, not the image itself. You want the URL behind the [Download] button.

ah ok thanks I will check

**Comment by Murmele** [comment]
bot, build com.github.chipmuenk.pyfda

**Comment by Murmele** [comment]
Was tested also by the original author and he said it works without any issues

File: com.github.chipmuenk.pyfda.yaml [hunk 11]
```diff
@@ -0,0 +1,373 @@
...
+command: pyfdax
+rename-icon: pyfda_icon #Image will renamed to match the app-id konvention
+rename-appdata-file: pyfda.appdata.xml
+rename-desktop-file: pyfda.desktop
+finish-args:
+- "--share=ipc"
+- "--socket=wayland"
+- "--socket=fallback-x11"
+- "--device=dri"
+- "--own-name=com.github.chipmuenk.pyfda"
```
Comment (hfiguiere) [inline com.github.chipmuenk.pyfda.yaml:16]: this is implicitely granted based on the app-id

```suggestion
```

File: com.github.chipmuenk.pyfda.yaml [hunk 12]
```diff
@@ -0,0 +1,373 @@
...
+- "--filesystem=home"
```
Comment (hfiguiere) [inline com.github.chipmuenk.pyfda.yaml:17]: does it really need home filesystem access.

File: matplotlib-setup.cfg [hunk 13]
```diff
@@ -0,0 +1,43 @@
+# Rename this file to setup.cfg to modify Matplotlib's build options.
```
Comment (hfiguiere) [inline matplotlib-setup.cfg:1]: is this needed?

File: com.github.chipmuenk.pyfda.yaml [hunk 14]
```diff
@@ -0,0 +1,373 @@
+app-id: com.github.chipmuenk.pyfda
+runtime: org.kde.Platform
+runtime-version: '5.15-21.08'
+sdk: org.kde.Sdk
+# when the runtime version changes, check if python version changed, because in pyqt5 the
+# python version is hardcoded!
+command: pyfdax
+rename-icon: pyfda_icon #Image will renamed to match the app-id konvention
```
Comment (hfiguiere) [inline com.github.chipmuenk.pyfda.yaml:8]:
```suggestion
rename-icon: pyfda_icon #Image will renamed to match the app-id convention
```

**Review by hfiguiere — CHANGES_REQUESTED** [review]

File: com.github.chipmuenk.pyfda.yaml (hunk 12 above)
Comment (chipmuenk) [reply com.github.chipmuenk.pyfda.yaml:17]: Currently I'm using a `~/.pyfda` directory for configuration files that is created during the first startup. For a `pip install pyfda` this shouldn't be a problem, but for flatpak installation I'm not happy with the broad access rights to do this either. I'm also not sure what "home filesystem access" means when you install the flatpak system-wide (i.e. with elevated rights). 

Do you have a suggestion how to circumvent this problem? I do need some read / write access right for importing / exporting filter designs etc.

File: matplotlib-setup.cfg (hunk 13 above)
Comment (Murmele) [reply matplotlib-setup.cfg]: I remove the command, I just copied the file and have overseen this command. It works without renaming, because it is defined in an environment variable

File: com.github.chipmuenk.pyfda.yaml (hunk 12 above)
Comment (Murmele) [reply com.github.chipmuenk.pyfda.yaml:16]: for the configuration file we might use one of the xdg dirs:
https://docs.flatpak.org/en/latest/flatpak-command-reference.html
maybe FLATPAK_CONFIG_DIR?

Comment (hfiguiere) [reply com.github.chipmuenk.pyfda.yaml:16]: `--persist=.pyfda` would persist the directory. until the app support xdg.

Comment (Murmele) [reply com.github.chipmuenk.pyfda.yaml:16]: I set persist

**Comment by Murmele** [comment]
bot build com.github.chipmuenk.pyfda

**Comment by chipmuenk** [comment]
@Murmele @hfiguiere The `--persist` option should do the job until I have added xdg support. "xdg" sounds like a good cross-platform concept to me that I didn't know about, unfortunately. If I understand it right, `--persist` maps `~/.pyfda` to a directory that is read/writable by the flathub app? Can the `.pyfda` directory be created if it doesn't exist during the first run of the app?

Can I detect whether pyfda is running in a flatpak context (currently, I'm checking whether pyfda is running within a pyinstaller package)?

Another problem would be to define an import / export directory for filter designs. Is it possible to "persist" e.g. the `~/Documents` directory (do all Linux distributions have / create this?).

I noticed that the app has internet access which is good as the help function links to `readthedocs.io` URLs.

Sorry for the lot of questions, this is my first contact with flatpak / flathub!

File: com.github.chipmuenk.pyfda.yaml [hunk 15]
```diff
@@ -0,0 +1,373 @@
...
+      - python3 setup.py install --prefix=${FLATPAK_DEST} --root=/ --optimize=1
+    sources:
+      - type: archive
+        url: https://files.pythonhosted.org/packages/06/97/56a6f56ce44578a69343449aa5a0d98eefe04085d69da539f3034e2cd5c1/xlwt-1.3.0.tar.gz
+        sha256: c59912717a9b28f1a3c2a98fd60741014b06b043936dcecbc113eaaada156c88
+        x-checker-data:
+          type: pypi
+          name: xlwt
+
+  - "python3-Pillow.json" # pyfda does not startup without
```
Comment (barthalion) [inline com.github.chipmuenk.pyfda.yaml:218]: What's the point of splitting this file?

File: com.github.chipmuenk.pyfda.yaml [hunk 16]
```diff
@@ -0,0 +1,373 @@
...
+      - pip3 install . --prefix=/app --no-index --find-links .
+    sources:
+      - type: archive
+        url: https://files.pythonhosted.org/packages/15/d1/d8798b83e953fd6f86ca9b50f93eec464a9305b0661469c8234e61095481/flit_core-3.7.1.tar.gz
+        sha256: 14955af340c43035dbfa96b5ee47407e377ee337f69e70f73064940d27d0a44f
+        x-checker-data:
+          type: pypi
+          name: flit_core
+          
+  - python3-tomli.json # needed by python-setuptools-scm-git-archive
```
Comment (barthalion) [inline com.github.chipmuenk.pyfda.yaml:103]: What's the point of splitting this file?

**Comment by barthalion** [comment]
bot build com.github.chipmuenk.pyfda

File: com.github.chipmuenk.pyfda.yaml (hunk 12 above)
Comment (barthalion) [reply com.github.chipmuenk.pyfda.yaml:16]: If you're passing `--filesystem=home`, `--persist` has no effect.

**Comment by barthalion** [comment]
bot, build com.github.chipmuenk.pyfda

**Comment by barthalion** [comment]
> The `--persist` option should do the job until I have added xdg support. "xdg" sounds like a good cross-platform concept to me that I didn't know about, unfortunately. If I understand it right, `--persist` maps `~/.pyfda` to a directory that is read/writable by the flathub app? Can the `.pyfda` directory be created if it doesn't exist during the first run of the app?

As I commented inline, `--persist` won't do anything if full access to home is granted.

> Can I detect whether pyfda is running in a flatpak context (currently, I'm checking whether pyfda is running within a pyinstaller package)?

The most reliable way is to check if the `/.flatpak-info` exists. Note that you don't need to read it, just checking if it's there is enough.

> Another problem would be to define an import / export directory for filter designs. Is it possible to "persist" e.g. the `~/Documents` directory (do all Linux distributions have / create this?).

The ideal situation is when the app uses filesystem portal. For Qt apps, you can read the docs at https://docs.flatpak.org/en/latest/portals-qt.html. This means the user chooses an arbitrary file to read from or save to, and it's safely exposed into the sandbox. 

To expand upon my `--persist` command, if you're fine with exposing a single directory, `--filesystem=xdg-documents` and `--persist=.pyfda` should work fine together.

> I noticed that the app has internet access which is good as the help function links to `readthedocs.io` URLs.

This is probably not needed then, as there's a portal for opening external URLs that will redirect it to web browser. If that's the only use for network accerss, I would drop it.

**Comment by chipmuenk** [comment]
Thanks for the detailled feedback!

- Full home access will be removed
- Providing read / write access to one directory via `--filesystem=xdg-documents` should be sufficient. Sorry if this is a stupid question but which path is this when viewed from the flatpak app and when viewed from the file system?
- There is no explicit rule allowing internet access for the app so I think the portal for opening external URLs is being used already

**Comment by Murmele** [comment]
bot, build com.github.chipmuenk.pyfda

**Comment by Murmele** [comment]
> Thanks for the detailled feedback!
> 
>     * Full home access will be removed
> 
>     * Providing read / write access to one directory via `--filesystem=xdg-documents` should be sufficient. Sorry if this is a stupid question but which path is this when viewed from the flatpak app and when viewed from the file system?
> 
>     * There is no explicit rule allowing internet access for the app so I think the portal for opening external URLs is being used already

I remove the home access and added access only to xdg-documents. So only the documents folder is accessible

**Comment by chipmuenk** [comment]
> I remove the home access and added access only to xdg-documents. So only the documents folder is accessible

> I found in the arch wiki:

> ~/.config/user-dirs.dirs
> XDG_DOCUMENTS_DIR="$HOME/Documents"

Yes, that looks good - thank you!

**Comment by barthalion** [comment]
It looks to me like pyfda correctly uses the filesystem portal right now, as it uses native GTK file chooser and allows me to choose any file from my file system. I'd just drop the `--filesystem=xdg-documents` line altogether.

Other than this, LGTM.

**Comment by chipmuenk** [comment]
Same here!

**Comment by Murmele** [comment]
Sorry I am on holidays until 19.04 i will change it after if not someone else cha ges

**Comment by barthalion** [comment]
bot, build com.github.chipmuenk.pyfda

**Comment by Murmele** [comment]
@barthalion is itpossible to add @chipmuenk as editor to the repository?

**Comment by barthalion** [comment]
Yes, this is already handled by my `/merge` comment above. @chipmuenk, please remember to accept the invitation GitHub has sent you, as it expires after some time.

**Comment by chipmuenk** [comment]
My 2FA doesn't work on my laptop at the moment, so I couldn't accept the invitation earlier. Thanks a lot for your support to all of you!
